rust/worker/src/segment/types.rs

```diff
@@ -136,131 +141,270 @@ pub struct MaterializedLogRecord<'referred_data> {
+}
+
+impl<'log_data, 'segment_data: 'log_data> HydratedMaterializedLogRecord<'log_data, 'segment_data> {
+    pub fn get_offset_id(&self) -> u32 {
+        self.materialized_log_record.offset_id
+    }
+
+    pub fn get_operation(&self) -> MaterializedLogOperation {
+        self.materialized_log_record.final_operation
+    }
+
+    pub fn get_user_id(&self) -> &'log_data str {
+        if let Some(id) = self.materialized_log_record.user_id_at_log_index {
+            return self.logs.get(id).unwrap().record.id.as_str();
+        }
+
+        if let Some(data_record) = self.segment_data_record.as_ref() {
+            data_record.id
+        } else {
+            panic!("Expected at least one source of user id")
         }
     }
 
-    // Performs a deep copy of the user id so only use it if really
-    // needed. If you only need reference then use merged_user_id_ref below.
-    pub(crate) fn merged_user_id(&self) -> String {
-        match self.user_id {
-            Some(id) => id.to_string(),
-            None => match &self.data_record {
-                Some(data_record) => data_record.id.to_string(),
-                None => panic!("Expected at least one user id to be set"),
-            },
+    pub fn document_ref_from_log(&self) -> Option<&'log_data str> {
+        match self.materialized_log_record.final_document_at_log_index {
+            Some(offset) => Some(self.logs.get(offset).unwrap().record.document.as_ref()?),
+            None => None,
         }
     }
 
-    pub(crate) fn merged_user_id_ref(&self) -> &str {
-        match self.user_id {
-            Some(id) => id,
-            None => match &self.data_record {
-                Some(data_record) => data_record.id,
-                None => panic!("Expected at least one user id to be set"),
-            },
+    pub fn document_ref_from_segment(&self) -> Option<&'segment_data str> {
+        self.segment_data_record
+            .as_ref()
+            .map(|data_record| data_record.document)?
+    }
+
+    pub fn merged_document_ref(&self) -> Option<&'log_data str> {
+        if self
+            .materialized_log_record
+            .final_document_at_log_index
+            .is_some()
+        {
+            return self.document_ref_from_log();
+        }
+
+        if self.materialized_log_record.final_operation
+            == MaterializedLogOperation::OverwriteExisting
+            || self.materialized_log_record.final_operation == MaterializedLogOperation::AddNew
+        {
+            None
+        } else {
+            self.document_ref_from_segment()
         }
     }
 
-    // Performs a deep copy of the metadata so only use it if really
-    // needed. If you only need reference then use merged_metadata_ref below.
-    pub(crate) fn merged_metadata(&self) -> HashMap<String, MetadataValue> {
+    /// Performs a deep copy of the metadata so only use this if really needed.
+    pub fn merged_metadata(&self) -> HashMap<String, MetadataValue> {
```

Comment:
We cannot have a `merged_metadata_ref()` method anymore. It would need to have this signature:
```suggestion
    pub fn merged_metadata_ref(&self) -> HashMap<&'log_data str, &'log_data MetadataValue> {
```
(semantically correct because `'segment_data` outlives `'log_data`). However, this is not possible.

Consider a simpler method:

```rust
fn log_metadata_ref(&self) -> HashMap<&'segment_data str, &'segment_data MetadataValue> {
    let mut final_metadata: HashMap<&str, &MetadataValue> = HashMap::new();

    if let Some(data_record) = &self.get_data_record() {
        if let Some(meta) = &data_record.metadata {
            for (meta_key, meta_val) in meta {
                final_metadata.insert(meta_key, meta_val);
            }
        }
    }

    final_metadata
}
```

This does not compile because it returns data borrowed from `self` with an elided lifetime instead of the `'segment_data` lifetime. The lifetime is tied to `self` because `self.segment_data_record` (unlike other fields) is owned rather than borrowed.

`HydratedMaterializedLogRecord.segment_data_record` cannot be borrowed because something must own the `DataRecord`. We want to borrow it for the duration of the reader lifetime, but the blockfile reader returns `DataRecord<'_>`, not `&DataRecord<'_>`.